pkg/apply/destroyer.go

```diff
@@ -68,107 +80,62 @@ func (d *Destroyer) Initialize() error {
 // happens asynchronously on progress and any errors are reported
 // back on the event channel.
 func (d *Destroyer) Run(inv inventory.InventoryInfo, option *DestroyerOption) <-chan event.Event {
-	ch := make(chan event.Event)
-
+	eventChannel := make(chan event.Event)
+	d.invClient.SetDryRunStrategy(option.DryRunStrategy)
 	go func() {
-		defer close(ch)
-		// TODO: We should see if we can avoid having the dry-run strategy
-		// as a field on the inventory client.
-		d.invClient.SetDryRunStrategy(option.DryRunStrategy)
-
-		// Start the event transformer goroutine so we can transform
-		// the Prune events emitted from the Prune function to Delete
-		// Events. That we use Prune to implement destroy is an
-		// implementation detail and the events should not be Prune events.
-		tempChannel, completedChannel := runPruneEventTransformer(ch)
-		taskContext := taskrunner.NewTaskContext(tempChannel)
-		err := d.PruneOptions.Prune(inv, nil, sets.NewString(), taskContext, prune.Options{
-			DryRunStrategy:    option.DryRunStrategy,
-			PropagationPolicy: metav1.DeletePropagationBackground,
-			InventoryPolicy:   option.InventoryPolicy,
-		})
+		defer close(eventChannel)
+		mapper, err := d.provider.Factory().ToRESTMapper()
 		if err != nil {
-			ch <- event.Event{
-				Type: event.ErrorType,
-				ErrorEvent: event.ErrorEvent{
-					Err: errors.WrapPrefix(err, "error pruning resources in cluster", 1),
-				},
-			}
+			handleError(eventChannel, err)
 			return
 		}
-
-		// Now delete the inventory object as well.
-		err = d.invClient.DeleteInventoryObj(inv)
+		// Retrieves the ids for the resources in the cluster from the inventory.
+		pruneIds, err := d.invClient.GetClusterObjs(inv)
 		if err != nil {
-			ch <- event.Event{
-				Type: event.ErrorType,
-				ErrorEvent: event.ErrorEvent{
-					Err: errors.WrapPrefix(err, "error deleting inventory object", 1),
-				},
-			}
+			handleError(eventChannel, err)
 			return
 		}
-
-		// Close the tempChannel to signal to the event transformer that
-		// it should terminate.
-		close(tempChannel)
-		// Wait for the event transformer to complete processing all
-		// events and shut down before we continue.
-		<-completedChannel
-		ch <- event.Event{
-			Type: event.ActionGroupType,
-			ActionGroupEvent: event.ActionGroupEvent{
-				GroupName: "destroyer",
-				Type:      event.Finished,
-				Action:    event.DeleteAction,
+		resourceObjs := &ResourceObjects{LocalInv: inv, PruneIds: pruneIds}
+		klog.V(4).Infoln("destroyer building task queue...")
+		taskBuilder := &solver.TaskQueueBuilder{
+			PruneOptions: d.PruneOptions,
```

Comment:
It seems strange here that `PruneOptions` are provided to the `TaskBuilder`. Why aren't this just part of the `solver.Options`?

Reply:
`TaskQueueSolver` was renamed to `TaskQueueBuilder`, but I did not change this. `PruneOptions` was a parameter to `TaskQueueSolver`. If we want to change this, let's do it in another smaller PR.